airbyte-config/init/src/main/java/io/airbyte/config/init/RemoteConnectorCatalogPersistence.java

```diff
@@ -0,0 +1,225 @@
+      LOGGER.warn(
+          "Unable to retrieve the remote connector catalog. Using the catalog bundled with Airbyte. This warning is expected if this Airbyte cluster does not have internet access.",
+          e);
+      this.connectorCatalog = ImmutableMap.<AirbyteConfig, Map<String, JsonNode>>builder()
+          .put(ConfigSchema.STANDARD_SOURCE_DEFINITION, Collections.emptyMap())
+          .put(ConfigSchema.STANDARD_DESTINATION_DEFINITION, Collections.emptyMap())
+          .build();
+    }
+  }
+
+  private static JsonNode getRemoteConnectorCatalog(URI catalogUrl, Duration timeout) throws IOException, InterruptedException {
+    final HttpRequest request = HttpRequest.newBuilder(catalogUrl).timeout(timeout).header("accept", "application/json").build();
+    final HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+    if (response.statusCode() >= 400) {
+      throw new IOException(
+          "getRemoteConnectorCatalog request ran into status code error: " + response.statusCode() + "with message: " + response.getClass());
+    }
+    try {
+      return Jsons.deserialize(response.body());
+    } catch (final RuntimeException e) {
+      throw new IOException("Could not deserialize JSON response: ", e);
+    }
+
+  }
+
+  private ImmutableMap<AirbyteConfig, Map<String, JsonNode>> parseConnectorCatalog(JsonNode catalog) {
+    return ImmutableMap.<AirbyteConfig, Map<String, JsonNode>>builder()
+        .put(ConfigSchema.STANDARD_SOURCE_DEFINITION, parseRemoteConfigs(catalog, ConfigSchema.STANDARD_SOURCE_DEFINITION))
+        .put(ConfigSchema.STANDARD_DESTINATION_DEFINITION, parseRemoteConfigs(catalog, ConfigSchema.STANDARD_DESTINATION_DEFINITION))
+        .build();
+
+  }
+
+  private static Map<String, JsonNode> parseRemoteConfigs(final JsonNode rawConnectorCatalog, final ConfigSchema configType) {
+    final String configKey = CONFIG_SCHEMA_CONNECTOR_CATALOG_MAP.get(configType);
+    final JsonNode rawConfigs = rawConnectorCatalog.get(configKey);
+    if (rawConfigs == null) {
+      return Collections.emptyMap();
+    }
+
+    final List<JsonNode> configKeyElements = MoreIterators.toList(rawConfigs.elements());
+
+    final Map<String, JsonNode> configs = configKeyElements.stream().collect(Collectors.toMap(
+        json -> json.get(configType.getIdFieldName()).asText(),
+        json -> json));
+    final Map<String, JsonNode> configsWithMissingField = configs.entrySet().stream()
+        .collect(Collectors.toMap(Entry::getKey, e -> {
+          final JsonNode output =
+              addMissingCustomField(
+                  addMissingPublicField(
+                      addMissingTombstoneField(e.getValue())));
+          AirbyteConfigValidator.AIRBYTE_CONFIG_VALIDATOR.ensureAsRuntime(configType, output);
+          return output;
+        }));
+    return configsWithMissingField;
+  }
+
+  private static JsonNode addMissingTombstoneField(final JsonNode definitionJson) {
```

Comment:
This and the ones below look like something that should be shared with the current implementation.